cpp/merge_differential_and_full_marc_updates.cc

```diff
@@ -374,54 +369,213 @@ void AppendFileOrDie(const std::string &append_target, const std::string &append
+const std::string EXTRACT_AND_APPEND_SCRIPT("extract_IDs_in_erase_format.sh");
+
+
+void ExtractAndAppendIDs(const std::string &marc_filename, const std::string &deletion_list_filename) {
+    if (unlikely(ExecUtil::Exec(EXTRACT_AND_APPEND_SCRIPT, { marc_filename, deletion_list_filename }) != 0))
+	LogSendEmailAndDie("\"" + EXTRACT_AND_APPEND_SCRIPT + "\" with arguments \"" + marc_filename + "\" and \""
+			   + deletion_list_filename + "\" failed!");
+}
+
+
+/** \brief If "old_name" is non-empty, rename it to "new_name", o/w create an empty file named "new_name". */
+void MoveOrCreateFileOrDie(const std::string &old_name, const std::string &new_name) {
+    if (old_name.empty()) {
+	if (unlikely(not FileUtil::MakeEmpty(new_name)))
+	    LogSendEmailAndDie("failed to create an empty file named \"" + new_name + "\"! ("
+			       + std::string(::strerror(errno)) + ")");
+    } else if (unlikely(not FileUtil::RenameFile(old_name, new_name, /* remove_target = */ true))) {
+	    LogSendEmailAndDie("failed to rename \"" + old_name + "\" to \"" + new_name + "\"! ("
+			       + std::string(::strerror(errno)) + ")");
+    }
+}
+
+
+/** \brief Replaces "filename"'s ending "old_suffix" with "new_suffix".
+ *  \note Aborts if "filename" does not end with "old_suffix".
+ */
+std::string ReplaceSuffix(const std::string &filename, const std::string &old_suffix, const std::string &new_suffix) {
+    if (unlikely(not StringUtil::EndsWith(filename, old_suffix)))
+	LogSendEmailAndDie("in ReplaceSuffix: \"" + filename + "\" does not end with \"" + old_suffix + "\"!");
+    return filename.substr(0, filename.length() - old_suffix.length()) + new_suffix;
+}
+
+
+const std::string LOCAL_DELETION_LIST_FILENAME("deletions.list");
+
+
+void ApplyUpdate(const unsigned apply_count, const std::string &deletion_list_filename, const std::string &differential_archive) {
+    MoveOrCreateFileOrDie(deletion_list_filename, LOCAL_DELETION_LIST_FILENAME);
+
+    // Unpack the differential archive and extract control numbers from its members appending them to the
+    // deletion list file:
+    if (not differential_archive.empty()) {
+	Log("updating the deletion list based on control numbers found in the files contained in the differential MARC archive.");
+	std::vector<std::string> extracted_names;
+	ExtractMarcFilesFromArchive(differential_archive, &extracted_names, "diff_");
+	for (const auto &extracted_name : extracted_names)
+	    ExtractAndAppendIDs(extracted_name, LOCAL_DELETION_LIST_FILENAME);
+    }
+
+    const std::string old_name_suffix("." + std::to_string(apply_count - 1));
+    std::string title_marc_basename, superior_marc_basename, normdata_marc_basename;
+    GetBasenamesOrDie(&title_marc_basename, &superior_marc_basename, &normdata_marc_basename, old_name_suffix);
+
+    const std::string new_name_suffix("." + std::to_string(apply_count));
+    std::string diff_filename;
+
+    // Update the title data:
+    std::string diff_filename_pattern("diff_.*a001.raw");
```

Comment:
Files with a001.raw and b001.raw need to be combined in to a single a001.raw file